### PR TITLE
Replace regex with `winnow` based parser

### LINKED ISSUE #73: Make `Procfile` validation more strict
Both this buildpack and Codon will happily accept a `Procfile` with contents:

```
web
```

And will set the command to the empty string.

Should we make both this buildpack and Codon reject this?

See:
https://github.com/heroku/procfile-cnb/blob/683de9b9877957ca0a5e6ba6733059cb55b5c4d9/src/procfile.rs#L41-L61

### LINKED ISSUE #178: Stop using regex for validation
This buildpack currently uses regex to validate Procfile contents:
https://github.com/heroku/procfile-cnb/blob/a1ee3659a5a95df96bf7b41a5f9dffb40678f4dd/src/procfile.rs#L38-L43

This causes quite a bit of buildpack bloat, particular with newer versions of the `regex` crate, which have gotten even larger (since by default it favours runtime performance for large/complex payloads/regex over compile time/binary size) - see https://github.com/heroku/procfile-cnb/pull/175#issuecomment-1686312268.

The original reason for using regex was likely to ensure compatibility with the existing Codon Procfile parsing, however, (a) it would be easy to maintain that even without regex, (b) we've talked about wanting more strict parsing anyway (#73).

### LINKED ISSUE #251: Enforce RFC 1123 on procfile process types
Kubernetes restricts some resource names to [RFC 1123](https://datatracker.ietf.org/doc/html/rfc1123), requiring that names match rules for subdomains. This buildpack produces process types that may be inconsistent with this spec. Notable inconsistencies:

- This buildpack supports capital letters (e.g. `Web`)
- This buildpack supports underscores (e.g. `my_worker`)
- This buildpack supports non-alphanumerics at the start or end of the string (e.g. `worker2`, `2nd-worker`)

Process types like these may cause issues when running the build result in a k8s environment, resulting in errors like these:

```
a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

We could:
- Coerce the entries to proper casing (downcasing)
- Coerce the entries to use dashes (`gsub _ -`)
- Write warning messages for non-compliant naming (e.g. "The `fast_worker` process has an underscore. This may cause problems in kubernetes environments. Consider changing this to `fast-worker`.")
- Throw errors and fail the build for non-compliant naming

### PR DESCRIPTION
**Pull request by schneems**
Replace the existing `Regex` based "parser" with an actual honest-to-goodness parser/combinator based parser.

- Close #251
- Close #178
- Close #73

## How to review

Some suggestions on how to review this:

- Proofread/comment on SPEC.md in the fist commit.  This is the basis for the rest of the parser work so if you disagree with something there, it's not worth continuing on into the implementation
- Read the changes to the integration tests on the final result: Warnings have been added, but otherwise assert that it's a refactor.
- Read the added parsing error message in error.rs
- Read the unit tests in `procfile.rs`. Some of them were useful for development, the majority are asserting overall behavior and edge cases
- If you want to review the actual winnow and parser code, please book a pairing session with me. It's a lot and is better to talk through syncronously. Feel free to suggest any documentation typos etc.

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  - A line MAY contain a key/value pair where the key represents the name of a process and the value represents a command
+- Key
+  - A key's first and last character MUST be a lowercase alphanumeric (a-z0-9) character (but not `-`).
+  - All other key (middle) characters MUST be lowercase alphanumeric (a-z0-9) characters or hyphen `-`.
+  - Key length MUST be within the range `1..=63`
+  - An implementation MAY accept `_` as a middle character provided it converts it to `-` and issues a warning.
+  - An implementation MAY accept an uppercase character provided it is converted to lowercase characters and issues a warning.
+  - A key MAY begin with zero or more spaces provided they are not included in the return key.
+  - A colon MUST terminate a key `:` (the colon is not included in the key return value).
+- Value
```
Comment (joshwlewis) [inline SPEC.md:38]: I'd suggest a `Separator` entry. Separator would include `:` and any whitespace around it? That would prevent `:` from being part of the key and ` ` from being part of the value.

File: SPEC.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# CNB Procfile format specification
```
Comment (joshwlewis) [inline SPEC.md:1]: I like that this exists, but I wonder if this spec should be shared/enforced more widely?

File: SPEC.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  - A key's first and last character MUST be a lowercase alphanumeric (a-z0-9) character (but not `-`).
+  - All other key (middle) characters MUST be lowercase alphanumeric (a-z0-9) characters or hyphen `-`.
+  - Key length MUST be within the range `1..=63`
+  - An implementation MAY accept `_` as a middle character provided it converts it to `-` and issues a warning.
+  - An implementation MAY accept an uppercase character provided it is converted to lowercase characters and issues a warning.
+  - A key MAY begin with zero or more spaces provided they are not included in the return key.
+  - A colon MUST terminate a key `:` (the colon is not included in the key return value).
+- Value
+  - A value MUST be proceeded by a key.
+  - A value MAY begin with 0 or more spaces.
```
[the remark this answers is not in the document]
Comment (schneems) [reply SPEC.md:40]: Starting spaces are trimmed, we could memorialize that in the spec with:

```suggestion
  - A value MAY be preceded with 0 or more spaces. These spaces are not part of the value.
```

Trailing spaces are not trimmed, they're considered to be part of the value. I don't have a good use case for that behavior, but it's what we're currently doing and IDK if there's a world where whitespace might somehow be significant.

[the remark this answers is not in the document]
Comment (schneems) [reply SPEC.md:40]: I'm playing with using "separated" wording as well based on your other suggestion. Working up a stacked PR.

File: SPEC.md (hunk 1 above)
Comment (schneems) [reply SPEC.md:38]: Here's a stacked PR with that alternative https://github.com/heroku/buildpacks-procfile/pull/256. I like your suggestion.

File: SPEC.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (schneems) [reply SPEC.md:40]: Alternative wording in a stacked PR https://github.com/heroku/buildpacks-procfile/pull/256

File: SPEC.md (hunk 2 above)
Comment (schneems) [reply SPEC.md:1]: > shared

You mean you're wanting more feedback from a diverse group of people (such as other Heroku teams) or you're wanting to promote the format to a larger group outside of Heroku such as language communities (or both)?

I'm open to both. I'm also open to suggestions. Perhaps we link to the SPEC from the readme and make it more prominent.

> enforced

We could share the logic by providing a stand-alone parser utility that converts it to some other format like json or toml. We could start by calling both parsers in cedar and instrumenting failures. While people are using `#` like comments in today's procfile, technically any leading character would work `//`.

I also had the idea that I could expose the parser or ship it via crates.io so that other buildpacks could use it. Some buildpacks have logic like "If app has a Procfile..." but it would be more robust to share the same parsing logic.

Getting a little ahead of myself. We could also consider upstreaming RFC 1123 to the CNB spec to make it more clear so then it's not "the procfile buildpack supports the intersection of kubernetes and CNB" and then becomes simply "the procfile buildpack supports the CNB spec".

Comment (joshwlewis) [reply SPEC.md:1]: Oh, I was just thinking share/enforce this across Heroku more broadly. This will be the third Procfile parser at Heroku, and I doubt they are consistent. This isn't a blocker for this PR though.

I'd support adjusting the CNB spec to have process names that match RFC 1123 if you wanted to do that. I expect many / most CNB users expect to use apps built with CNBs in kubernetes in some way.

File: SPEC.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+- Comments
+  - A line MAY contain a comment
+  - Comments MUST contain a `#` as the first visible character on each line.
+  - A comment MAY be proceeded by one or more spaces.
+- Empty line
+  - A Procfile MAY contain empty lines
+  - An empty line MUST be zero or more spaces followed by a line ending or end of file (EOF).
+- Key/Value pairs
+  - A line MAY contain a key/value pair where the key represents the name of a process and the value represents a command
+  - A key MUST be separated from its value by a colon (`:`) followed by zero or more spaces.
```
Comment (joshwlewis) [inline SPEC.md:30]: I don't think this spec covers a space before the `:`? Like `my-process : bin/blah`. I think that should be legal, but I could be wrong.

**Review by joshwlewis — APPROVED** [review]
This probably warrants a major version bump. Looks good to me!

File: SPEC.md (hunk 4 above)
Comment (schneems) [reply SPEC.md:30]: > I think that should be legal, but I could be wrong.

Looking at the prior regex

```
^[[:space:]]*([a-zA-Z0-9_-]+):?\\s+(.*)[[:space:]]*
```

- starts with one or more whitespace characters
- alphanumeric or `_` or `-`
- funny enough the `:` is optional, but it doesn't support spaces before `:`
- A space after the (optional) `:` is enforced (which is a requirement we've loosened up). It cannot be a tab, it must be a space (which we've loosened up).
- Then it grabs the value
- Then the regex makes it look like it strips off any trailing spaces but it doesn't. Rust won't strip it out:

image

Nor will ruby:

```
irb(main):005:0> match = "web: rails s         ".match(/^[[:space:]]*([a-zA-Z0-9_-]+):?\s+(.*)[[:space:]]*/);
=> #<MatchData "web: rails s         " 1:"web" 2:"rails s         ">
irb(main):006:0> match[2]
=> "rails s         "
```

So a space before the `:` wouldn't work today, but instead of an error or warning you'll quietly get an ignored line. I don't think we want to support optional `:`.

Comment (edmorley) [reply SPEC.md:30]: I'm also in favour of making the spec (and the Procfile CNB implementation) more strict than the current implementation in cases where we think it makes sense. So long as the Procfile CNB error message is clear, users can tweak their Procfile if needed as part of migrating to CNB.

Comment (Malax) [reply SPEC.md:30]: +1 on @edmorley's comment. I imagine most of the allowed weird `Procfile`s on cedar are accidentally valid.
